plugins/woocommerce/src/Internal/Admin/Orders/PostsRedirectionController.php

```diff
@@ -0,0 +1,160 @@
+<?php
+namespace Automattic\WooCommerce\Internal\Admin\Orders;
+
+use Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController;
+
+/**
+ * When {@see OrdersTableDataStore} is in use, this class takes care of redirecting admins from CPT-based URLs
+ * to the new ones.
+ */
+class PostsRedirectionController {
+
+	/**
+	 * Instance of the PageController class.
+	 *
+	 * @var PageController
+	 */
+	private $page_controller;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param PageController $page_controller Page controller instance. Used to generate links/URLs.
+	 */
+	public function __construct( PageController $page_controller ) {
+		$this->page_controller = $page_controller;
+
+		if ( ! wc_get_container()->get( CustomOrdersTableController::class )->custom_orders_table_usage_is_enabled() ) {
+			return;
+		}
+
+		add_action(
+			'load-edit.php',
+			function() {
+				$this->maybe_redirect_to_orders_page();
+			}
+		);
+
+		add_action(
+			'load-post-new.php',
+			function() {
+				$this->maybe_redirect_to_new_order_page();
+			}
+		);
+
+		add_action(
+			'load-post.php',
+			function() {
+				$this->maybe_redirect_to_edit_order_page();
+			}
+		);
+	}
+
+	/**
+	 * If needed, performs a redirection to the main orders page.
+	 *
+	 * @return void
+	 */
+	private function maybe_redirect_to_orders_page(): void {
+		if ( ! isset( $_GET['post_type'] ) || 'shop_order' !== $_GET['post_type'] ) {
+			return;
+		}
+
+		// Respect query args, except for 'post_type'.
+		$query_args = wp_unslash( $_GET );
+		unset( $query_args['post_type'] );
+
+		// Remap 'post_status' arg.
+		if ( isset( $query_args['post_status'] ) ) {
+			$query_args['status'] = $query_args['post_status'];
+			unset( $query_args['post_status'] );
+		}
+
+		$new_url = $this->page_controller->get_orders_url();
+		$new_url = add_query_arg( $query_args, $new_url );
+
+		wp_safe_redirect( $new_url, 301 );
+		exit;
+	}
+
+	/**
+	 * If needed, performs a redirection to the new order page.
+	 *
+	 * @return void
+	 */
+	private function maybe_redirect_to_new_order_page(): void {
+		if ( ! isset( $_GET['post_type'] ) || 'shop_order' !== $_GET['post_type'] ) {
+			return;
+		}
+
+		// Respect query args, except for 'post_type'.
+		$query_args = wp_unslash( $_GET );
+		unset( $query_args['post_type'] );
+
+		$new_url = $this->page_controller->get_new_page_url();
+		$new_url = add_query_arg( $query_args, $new_url );
+
+		wp_safe_redirect( $new_url );
+		exit;
+	}
+
+	/**
+	 * If needed, performs a redirection to the edit order page.
+	 *
+	 * @return void
+	 */
+	private function maybe_redirect_to_edit_order_page(): void {
+		$post_id = absint( $_GET['post'] ?? 0 );
+
+		if ( ! $post_id || 'shop_order' !== get_post_type( $post_id ) || ! isset( $_GET['action'] ) ) {
```

Comment:
Should we also test for posts of type `shop_order_placehold`?

(I'm thinking here of cases where some kind of third party app generates these URLs based on its knowledge of order IDs, rather than cases where a human user captured and preserved a historic link.)

Reply:
Yes. That definitely makes sense.